<center> <h2> DS 3000 - Fall 2021</h2> </center>
<center> <h3> DS Report </h3> </center>


<center> <h3> Project Title</h3> </center>
<center><h4>Team Member Names</h4></center>


<hr style="height:2px; border:none; color:black; background-color:black;">

#### Executive Summary:

Add your summary here (100-150 words)

Provide a brief summary of your project. After reading this executive summary, your readers should have a rough understanding of what you did in this project. You can think of this summary in terms of the four sections of the report and write 1-2 sentences describing each section.



<hr style="height:2px; border:none; color:black; background-color:black;">

## Outline
1. <a href='#1'>INTRODUCTION</a>
2. <a href='#2'>METHOD</a>
3. <a href='#3'>RESULTS</a>
4. <a href='#4'>DISCUSSION</a>

<a id="1"></a>
<hr style="height:2px; border:none; color:black; background-color:black;">

## 1. INTRODUCTION

In this section, orient your readers to your project. You've already written some of these in previous deliverables. Based on your final analysis, revise your problem statement and write a concise introduction section. This section should touch upon the following points, but should be written in full paragraphs. Your writing should incorporate all of these points (and more if you like) in a coherent way. Remember that you are trying to convince your readers that this is an important problem to tackle. 

<h4>Problem Statement</h4>

* Describe the problem you would like to tackle. 
* What is the topic of your project? 
* What do you want to learn about it?

<h4>Significance of the Problem</h4>

* Why is it important to tackle this problem in your project?
* In what ways could the insights from this project be useful?
* **(optional)** Has there been previous work on your topic in terms of applying ML techniques to analyze similar datasets? Do some research into your topic. What algorithms were used? What was the performance of those algorithms? Cite your sources appropriately. You can use the numbered reference format or APA (if you are more comfortable with it).

<h4>Questions</h4>

* End this section with a list of data analysis questions you tackled in this project
* You should tie these questions to the problem statement and its significance
    * e.g. Given the aforementioned problem and its importance, we set out to tackle the following questions:
    
    * **Requirement:**
        * You should have at least one question tapping into the comparison of various machine learning algorithms (at least three) in predicting/classifying your target variable from your features variables.



While this section provides bullet points as prompts, your Introduction should consist of coherent paragraphs. You are required delete all the prompts from this template, except for headings. Failing to do so will result in substantial point deductions.

<a id="2"></a>
<hr style="height:2px; border:none; color:black; background-color:black;">

## 2. METHOD

### 2.1. Data Acquisition

* Describe where you obtained your data. Provide a link to the original source. 
* If you scraped your data, include your code as a **separate** Notebook file.
* Your data should be stored in an online repository (e.g., GitHub) and your code should retrieve your data from that online resource. You can read csv files from the Web in the same way that you read files from local drive.
* Describe the dataset (i.e., what it is about) and the number of variables/samples included.


In [1]:
# IMPORTS
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns

In [6]:
url = 'https://raw.githubusercontent.com/jjz17/NBA-MIP-Prediction-Model/main/all_player_seasons_outcome.csv'
data = pd.read_csv(url, index_col=0)
data

Player    Pos  Age   Tm   G  GS    MP   FG   FGA  \
Rk                                                                             
1     Mahmoud Abdul-Rauf\abdulma02     PG   31  VAN  41   0  11.9  2.9   6.0   
2      Tariq Abdul-Wahad\abdulta01     SG   26  DEN  29  12  14.5  1.5   3.8   
3    Shareef Abdur-Rahim\abdursh01     SF   24  VAN  81  81  40.0  7.5  15.8   
4         Cory Alexander\alexaco01     PG   27  ORL  26   0   8.7  0.7   2.2   
5     Courtney Alexander\alexaco02     PG   23  TOT  65  24  21.3  3.7   8.8   
..                             ...    ...  ...  ...  ..  ..   ...  ...   ...   
536         Delon Wright\wrighde01  SG-PG   28  TOT  63  39  27.7  3.8   8.2   
537       Thaddeus Young\youngth01     PF   32  CHI  68  23  24.3  5.4   9.7   
538           Trae Young\youngtr01     PG   22  ATL  63  63  33.7  7.7  17.7   
539          Cody Zeller\zelleco01      C   28  CHO  48  21  20.9  3.8   6.8   
540          Ivica Zubac\zubaciv01      C   23  LAC  72  33  22.3  3.6   5.5   

       FG%  ...  DRB  TRB  AST  STL  BLK  TOV   PF   PTS  Season  Outcome  
Rk          ...                                                            
1    0.488  ...  0.5  0.6  1.9  0.2  0.0  0.6  1.2   6.5    2000        0  
2    0.387  ...  1.6  2.0  0.8  0.5  0.4  1.2  1.9   3.8    2000        0  
3    0.472  ...  6.9  9.1  3.1  1.1  1.0  2.9  2.9  20.5    2000        0  
4    0.321  ...  1.0  1.0  1.4  0.6  0.0  1.0  1.1   2.0    2000        0  
5    0.417  ...  1.6  2.2  1.0  0.7  0.1  1.2  2.1   9.5    2000        0  
..     ...  ...  ...  ...  ...  ...  ...  ...  ...   ...     ...      ...  
536  0.463  ...  3.2  4.3  4.4  1.6  0.5  1.3  1.2  10.2    2020        0  
537  0.559  ...  3.8  6.2  4.3  1.1  0.6  2.0  2.2  12.1    2020        0  
538  0.438  ...  3.3  3.9  9.4  0.8  0.2  4.1  1.8  25.3    2020        0  
539  0.559  ...  4.4  6.8  1.8  0.6  0.4  1.1  2.5   9.4    2020        0  
540  0.652  ...  4.6  7.2  1.3  0.3  0.9  1.1  2.6   9.0    2020        0  

[9943 rows x 31 columns]

In [8]:
data.shape

(9943, 31)

## Description

### Player Data Per Season
The player data per season and Most-Improved-Pwas acquired from the *Basketball Reference* website and downloaded with the provided download links.


https://www.basketball-reference.com/awards/mip.html

https://www.basketball-reference.com/leagues/NBA_2021_per_game.html



The dataset contains an entry for each NBA player in the last 21 seasons (2000-2001 through 2020-2021).
\
It contains a plethora of statistics that describe the performance of each player during the given season.
\
There are 31 variables (columns) in the dataset which means there are 30 features and 1 target.
\
The dataset contains 9943 samples/entries.

### 2.2. Data Analysis
* Describe your predictive model in greater detail. What outcome variable are you going to predict from what feature variables? Why do you think those are important predictors?
* Describe why this is a supervised ML problem and identify the sub-category of the learning task (e.g. classification).
* What machine learning algorithms are you going to use? **Why?** You should list at least three algorithms. Provide a brief description of each algorithm and explain why you think they are appropriate for your data.

The outcome variable that we are going to predict is the MIP-Win-Shares variable, which describes the likelihood of a player to be awarded the Most-Improved-Player Award for the given season.
\
\
This is a supervised ML problem because we will eventually be giving the model the feature variables and corresponding outcome variable values in the training stage. The sub-category of our particular learning task is regression, as we are going to predict a continuous rather than a discrete variable.
\
\
We will be using 

<a id="3"></a>
<hr style="height:2px; border:none; color:black; background-color:black;">

## 3. RESULTS

### 3.1. Data Wrangling
In this section, you should do the following and explain why you are doing what you are doing. For each, you should include your code in a cell, followed by a sample output. For instance, if you are one-hot encoding one of your variables, you should first describe what it is and why you are doing it. You should then include your code in a cell, and the sample output should be available as well.

Do at least three of the following:
* Perform simple data cleaning (delete extra columns, deal with NA values, etc.)
* Perform data wrangling to get your features and target values (e.g., grouping your dataframe by columns, applying functions to format dataframes, etc.)
* Preprocess your variables (e.g., scaling/transforming feature variables to normalize them)
* Perform feature extraction (dummy variables, new features from existing features, etc.)

### 3.2. Data Exploration
* Generate appropriate data visualizations for your key variables identified in the previous section
* You should have at least three visualizations (and at least two different visualization types)
* For each visualization provide an explanation regarding the variables involved and an interpretation of the graph. In other words, what does the graph represent?
* Insert your visualizations as images as well (upload the graph images to an online source, e.g. github or imgbb, and embed those into the cells in Jupyter Notebook, similar to HWs). This is a requirement.


### 3.3. Model Training
* For your machine learning question(s), use the Training, Validation, and Testing approach through GridSearch
* Apply machine learning algorithms (apply at least three different algorithms)
* Train your algorithms using your **training set**

* Evaluate the performance of your algorithms using appropriate evaluation metrics
    * Use at least two different metrics
    * Remember GridSearch will use cross-validation

* Is there any indication of overfitting/underfitting?

* Use one feature selection technique to select a subset of your original features and train your models with the selected features as well (the idea is to see if you can achieve the same using a fewer number of features)

    * **Hint:** You can run your model with all your features and with your selected features, comparing the results (similar to the corresponding HW assignment).

### 3.4. Model Optimization
* Tune your models using hyperparameters appropriate for your algorithms
* Explain why you are doing this (e.g., to avoid overfitting, etc.)

### 3.5. Model Testing
* Test your tuned algorithms using your **testing set**

<a id="4"></a>
<hr style="height:2px; border:none; color:black; background-color:black;">

## 4. DISCUSSION
* Interpret your findings from 3.3, 3.4, and 3.5
    * Which algorithms did you compare?
    * Which algorithm(s) revealed best performance? With what parameters?
    * Which algorithm(s) should be used for your predictive model?
    * Based on your findings, can we use the features in your dataset to predict the outcome variable you identified using the algorithms you've applied? (It is okay if the answer is no. We're interested in the process, not the performance of the model.)

* Discuss the ethical implications of your project. Should your results be accepted at face value, why or why not? (e.g. any dataset bias or methodological issues?)

* End this section with a conclusion paragraph containing some pointers for future work
    * (e.g., get more data/features, perform another analysis, etc.)

<a id="5"></a>
<hr style="height:2px; border:none; color:black; background-color:black;">

### CONTRIBUTIONS
* Describe each team member's contributions to the report (who did what in each section)
* Remember this is a team effort!
* Each member of your team will later provide peer evaluation of other team members. Your final grade on the project will be based on those peer evaluations. A survey will be shared after the deadline for this deliverable.